base_layer/wallet/src/ffi.rs

```diff
@@ -120,284 +435,574 @@ pub unsafe extern "C" fn create_pending_transaction_outputs(
 ) -> bool
 {
-    // append output to tx
+    if wallet.is_null() {
+        return false;
+    }
+
+    if output.is_null() {
+        return false;
+    }
+
+    // (*wallet).pendingtransactionoutputs.addReceivedoutput((*output)) ? true : false;
+    return true;
 }
 
-/// Add a transaction kernel to a transaction struct
+/// Add an output to the wallet. `spent` is a boolean that indicates if this output is a spent or unspent output.
 #[no_mangle]
-pub unsafe extern "C" fn add_transaction_kernel(
-    transaction: *mut Transaction,
-    features: c_char, // KernelFeatures,
-    fee: c_ulonglong, // MicroTari,
-    lock_height: u64,
-    meta_info: *const c_char,     // Option<Byte[32]> - Option<HashOutput>,
-    linked_kernel: *const c_char, // Option<Byte[32]> - Option<HashOutput>,
-    excess: *const c_char,        // Byte[32] - Commitment,
-    excess_sig: *const c_char,    // Byte[32] - Signature,
+pub unsafe extern "C" fn wallet_add_pending_transaction_outputs(
+    wallet: *mut Wallet,
+    output: *mut PendingTransactionOutputs,
+    spent: bool,
 ) -> bool
 {
-    // append kernel to tx
+    if wallet.is_null() {
+        return false;
+    }
+
+    if output.is_null() {
+        return false;
+    }
+
+    match spent {
+        true => {},  //(*wallet).pendingtransactionoutputs.addSpentOutput((*output)) ? true : false;
+        false => {}, //(*wallet).pendingtransactionoutputs.addReceivedOutput((*output)) ? true : false;
+    }
+    return true;
+}
+
+/// TODO Methods to construct, free above 3 types
+
+#[no_mangle]
+pub unsafe extern "C" fn wallet_create_transaction(
+    inputs: *mut TransactionInputs,
+    outputs: *mut TransactionOutputs,
+    kernels: *mut TransactionKernels,
+    offset: *const PrivateKey,
+) -> *mut Transaction
+{
+    /// TODO null check
+    let t = Transaction::new(
+        (*inputs).0.clone(),
+        (*outputs).0.clone(),
+        (*kernels).0.clone(),
+        (*offset).clone(),
+    );
+    Box::into_raw(Box::new(t))
 }
 
 /// Add an completed transaction to the wallet.
 #[no_mangle]
-pub unsafe extern "C" fn add_transaction(wallet: *mut WalletFfi, pending_tx: *mut Transaction) -> bool {}
+pub unsafe extern "C" fn wallet_add_transaction(
+    wallet: *mut Wallet,
+    pending_tx: *mut Transaction,
+    inbound: bool,
+) -> bool
+{
+    return true;
+}
 
 /// Add a ReceivedTransactionProtocol instance to the wallet
 #[no_mangle]
-pub unsafe extern "C" fn add_pending_inbound_transaction(
-    wallet: *mut WalletFfi,
-    tx_id: c_ulonglong,               // u64,
-    public_spend_key: *const c_char,  // Byte[32] - PublicKey,
-    partial_signature: *const c_char, // Byte[32] - Signature,
-    // TransactionOutput
-    commitment: *const c_char, // Byte[32] - Commitment
-    proof: *const c_char,      // Byte[32] - Rangeproof
-    feature_flags: c_char,     // OutputFlags,
-    maturity: c_ulonglong,     // u64
+pub unsafe extern "C" fn wallet_add_pending_inbound_transaction(
```

Comment:
The Inbound, Outbound and Completed transactions are all different structures:
A pending Inbound Transaction is actually a ReceiverTransactionProtocol
A pending Outbound Transaction is actually a SenderTransactionProtocol
A completed Transaction is a Transaction.

In my original API I broke down the members of each of these items into the minimal set of primitives.